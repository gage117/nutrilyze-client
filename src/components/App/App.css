#App {
  background-color: #142850;
  min-height: 100vh;
}