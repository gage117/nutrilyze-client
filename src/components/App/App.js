import React from 'react'
import PrivateRoute from '../../Utils/PrivateRoute'
import PublicOnlyRoute from '../../Utils/PublicOnlyRoute'
import {BrowserRouter, Route, Switch } from 'react-router-dom'
import MainPage from '../../routes/MainPage/MainPage'
import LandingRoute from '../../routes/LandingRoute/LandingRoute'
import RegistrationPage from '../../routes/RegistrationPage/RegistrationPage'
import NotFoundPage from '../../routes/NotFoundPage/NotFoundPage'
import './App.css'
import LoginPage from '../../routes/LoginPage/LoginPage'
import authService from '../../services/auth-api-service'

class App extends React.Component {
  constructor(props) {
    super(props)
  
    this.state = {
       users: this.sortUsers([]),
       ingredients: this.sortIngredients([]),
       //meals: this.sortIngredients(this.props.Store.meals),
       error: null
    }
  }

  updateIngsFromFetch = (ingredientsArr) => {
    this.setState({
      ingredients: this.sortIngredients(ingredientsArr)
    })
  }

  updateUsersFromFetch = (user) => {
    this.setState({
      users: [user]
    })
  }

  setError = error => {
    console.error(error)
    this.setState({ error })
  }

  clearError = () => {
    this.setState({ error: null })
  }

  sortUsers = (array) => {
    return array.sort((a, b) => a.id - b.id)
  }

  sortIngredients = (array) => {
    function compare( a, b ) {
      if ( a.name < b.name ){
        return -1;
      }
      if ( a.name > b.name ){
        return 1;
      }
      return 0;
    }
    return array.sort(compare)
  }
  
  updateUserNutrition = (nutritionValues, usersToUpdate) => {
    let updatedUsers = [];
    /** Multiple user functionality, buggy, needs to only compare id not whole user */
    // updatedUsers = updatedUsers.filter(user => {
    //   return !usersToUpdate.includes(user)
    // })
    usersToUpdate.map(user => {
      const updatedUser = {
        id: user.id,
        name: user.name,
        user_name: user.user_name,
        calories: parseInt(user.calories) + parseInt(nutritionValues.calories),
        protein: parseFloat(parseFloat(user.protein) + parseFloat(nutritionValues.protein)).toFixed(2),
        carbs: parseFloat(parseFloat(user.carbs) + parseFloat(nutritionValues.carbs)).toFixed(2),
        sugar: parseFloat(parseFloat(user.sugar) + parseFloat(nutritionValues.sugar)).toFixed(2),
        fiber: parseFloat(parseFloat(user.fiber) + parseFloat(nutritionValues.fiber)).toFixed(2),
        fat: parseFloat(parseFloat(user.fat) + parseFloat(nutritionValues.fat)).toFixed(2),
        sodium: parseInt(user.sodium) + parseInt(nutritionValues.sodium),
      }
      authService.updateUser(updatedUser)
      updatedUsers.push(updatedUser)
      return this.sortUsers(updatedUsers)
    })
    
    this.setState({
      users: updatedUsers
    })
  }

  storeFood = (foodToStore) => {
    /** Checks if ingredient with same name already exists */
    let nameFound;
    const findFood = (food) => {
      if (food.name.toLowerCase() === foodToStore.name.toLowerCase()) {
        nameFound = true;
      }
    }
    this.state.ingredients.some(findFood)

    /** Adds food to store if nameFound isn't true */
    if (nameFound === true) {
      alert('Food with entered name already exists')
    } else {
      this.setState({
        ingredients: this.sortIngredients([...this.state.ingredients, foodToStore])
      })
    }
  }
  render() { 
    const handlers = {
      updateUserNutrition: this.updateUserNutrition,
      updateIngsFromFetch: this.updateIngsFromFetch,
      updateUsersFromFetch: this.updateUsersFromFetch,
      storeFood: this.storeFood,
      setError: this.setError,
      clearError: this.clearError

    }
    return (
      <div id="App">
        <BrowserRouter>
          <Switch>
            <Route
              exact
              path='/'
              component={LandingRoute}
            />
            <PublicOnlyRoute 
              exact
              path='/login'
              component={LoginPage}
            />
            <PublicOnlyRoute
              exact
              path='/register'
              component={RegistrationPage}
            />
            <PrivateRoute
              path='/user/:username'
              component={ 
                MainPage
              }
              props={{
                state: this.state,
                handlers: handlers
              }}
            />
            <Route
              component={NotFoundPage}
            />
          </Switch>
        </BrowserRouter>
      </div>
    )
  }
}

export default App
