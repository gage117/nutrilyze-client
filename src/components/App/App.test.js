import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';

describe('App.js', () => {
  const Store = {
    users: [
      {
        id: 1,
        name: 'Gage',
        calories: 666,
        carbs: 40,
        protein: 20,
        fat: 30,
        sodium: 20
      },
      {
        id: 2,
        name: 'Alanna',
        calories: 0,
        carbs: 0,
        protein: 0,
        fat: 0,
        sodium: 0
      }
    ],
    storedItems: [
      {
        name: 'Banana',
        calories: 0,
        carbs: 0,
        protein: 0,
        fat: 0,
        sodium: 0,
        serving_size: 0
      }
    ]
  }
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<App Store={Store}/>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<App Store={Store}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})