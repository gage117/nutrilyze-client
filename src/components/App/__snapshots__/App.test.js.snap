// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js renders the UI as expected 1`] = `
<div
  id="App"
>
  <nav
    className="Header"
  >
    <h1
      className="Header__h1"
    >
      <a
        className="Header__main-button"
        href="/"
        onClick={[Function]}
      >
        Nutrilyze
      </a>
    </h1>
    <div
      className="Header__not-logged-in"
    >
      <a
        className="Header__login-button Header__button"
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
      <a
        className="Header__register-button Header__button"
        href="/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
  </nav>
  <div
    id="landing-page"
  >
    <section
      className="lp__section"
    >
      <header
        className="lp__header"
      >
        <h3>
          Keep track of daily nutrition
        </h3>
      </header>
      <p>
        [
        <em>
          placeholder for screenshot of current-nutrition-info interface
        </em>
        ]
      </p>
      <p>
        Nutrilyze stores and tracks nutrition info of all foods you eat. Allowing easy counting, storing, and reading of your current and past intake. You can also add additional users to your profile, allowing simple tracking of multiple people.
      </p>
    </section>
    <section
      className="lp__section"
    >
      <header
        className="lp__header"
      >
        <h3>
          Store foods for easier intake tracking
        </h3>
      </header>
      <p>
        [
        <em>
          placeholder for screenshot of food-store interface
        </em>
        ]
      </p>
      <p>
        Saving the food along with it's nutritional value and serving size allows you to simply retrieve the item again later, keeping you from having to stare at the nutrition facts while you type it in every time.
      </p>
    </section>
    <section
      className="lp__section"
    >
      <header
        className="lp__header"
      >
        <h3>
          Combine ingredients into meals
        </h3>
      </header>
      <p>
        [
        <em>
          placeholder for screenshot of meals in UI
        </em>
        ]
      </p>
      <p>
        Storing ingredients and meals separately allows you to easily add up the total nutrients in a combination of ingredients you have around the house. You may have none, some, or all of these ingredients already stored and ready to combine. Any not already stored will be stored for the next meal that uses that ingredient!
      </p>
    </section>
    <section
      className="lp__section lp__last-section"
    >
      <header
        className="lp__header"
      >
        <p
          className="lp__footer"
        >
          Get started by
        </p>
        <a
          className="lp__link"
          href="/register"
          onClick={[Function]}
        >
          Registering
        </a>
        <p
          className="lp__footer"
        >
           or by using our 
        </p>
        <button
          className="lp__button"
          onClick={[Function]}
        >
          Demo Account
        </button>
      </header>
    </section>
  </div>
</div>
`;
